File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappUtils.java
package net.corda.plugins.cpk2;

import org.gradle.api.Action;
import org.gradle.api.InvalidUserDataException;
import org.gradle.api.Task;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.artifacts.ConfigurationContainer;
import org.gradle.api.artifacts.Dependency;
import org.gradle.api.artifacts.ResolvedArtifact;
import org.gradle.api.artifacts.ResolvedConfiguration;
import org.gradle.api.artifacts.ResolvedDependency;
import org.gradle.api.execution.TaskExecutionGraph;
import org.gradle.api.tasks.bundling.Jar;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.Element;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.StringReader;
import java.io.UncheckedIOException;
import java.lang.invoke.MethodHandles;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Collection;
import java.util.Enumeration;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringJoiner;
import java.util.function.Function;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.jar.Manifest;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static aQute.bnd.osgi.Constants.FIXUPMESSAGES_IS_DIRECTIVE;
import static aQute.bnd.osgi.Constants.FIXUPMESSAGES_IS_ERROR;
import static aQute.bnd.osgi.Constants.FIXUPMESSAGES_IS_WARNING;
import static aQute.bnd.osgi.Constants.FIXUPMESSAGES_RESTRICT_DIRECTIVE;
import static java.lang.invoke.MethodType.methodType;
import static java.util.Collections.max;
import static java.util.Collections.unmodifiableMap;
import static org.gradle.api.plugins.JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME;
import static org.gradle.api.plugins.JavaPlugin.JAR_TASK_NAME;

public final class CordappUtils {
    static final String CORDAPP_CPK_PLUGIN_ID = "net.corda.plugins.cordapp-cpk2";
    static final String CORDAPP_CONFIG_PLUGIN_ID = "net.corda.cordapp.cordapp-configuration";
    public static final String CORDAPP_TASK_GROUP = "Cordapp";

    static final String CORDA_API_GROUP = "net.corda";
    static final String ENTERPRISE_API_GROUP = "com.r3.corda";

    static final String CORDAPP_SEALING_SYSTEM_PROPERTY_NAME = "net.corda.cordapp.sealing.enabled";

    public static final String CPK_FILE_EXTENSION = "jar";

    static final String CORDAPP_CONFIGURATION_NAME = "cordapp";
    static final String CORDAPP_EXTERNAL_CONFIGURATION_NAME = "cordappExternal";
    static final String CORDAPP_PACKAGING_CONFIGURATION_NAME = "cordappPackaging";
    static final String CORDA_RUNTIME_ONLY_CONFIGURATION_NAME = "cordaRuntimeOnly";
    static final String CORDA_ALL_PROVIDED_CONFIGURATION_NAME = "cordaAllProvided";
    static final String CORDA_PRIVATE_CONFIGURATION_NAME = "cordaPrivateProvided";
    static final String CORDA_PROVIDED_CONFIGURATION_NAME = "cordaProvided";
    static final String CORDA_EMBEDDED_CONFIGURATION_NAME = "cordaEmbedded";
    public static final String ALL_CORDAPPS_CONFIGURATION_NAME = "allCordapps";
    static final String CORDA_CPK_CONFIGURATION_NAME = "cordaCPK";

    static final String CPK_DEPENDENCIES = "META-INF/CPKDependencies.json";

    // These tags are for the CPK file.
    static final String CPK_PLATFORM_VERSION = "Corda-CPK-Built-Platform-Version";
    public static final String CPK_CORDAPP_NAME = "Corda-CPK-Cordapp-Name";
    static final String CPK_CORDAPP_VERSION = "Corda-CPK-Cordapp-Version";
    static final String CPK_CORDAPP_LICENCE = "Corda-CPK-Cordapp-Licence";
    static final String CPK_CORDAPP_VENDOR = "Corda-CPK-Cordapp-Vendor";
    static final String CPK_FORMAT_TAG = "Corda-CPK-Format";
    static final String CPK_FORMAT = "2.0";

    // These tags are for the "main" JAR file.
    static final int PLATFORM_VERSION_X = 999;
    static final String IMPORT_POLICY_PACKAGES = "Import-Policy-Packages";
    static final String REQUIRED_PACKAGES = "Required-Packages";

    static final String CORDAPP_PLATFORM_VERSION = "Cordapp-Built-Platform-Version";
    static final String CORDAPP_CONTRACT_NAME = "Cordapp-Contract-Name";
    static final String CORDAPP_CONTRACT_VERSION = "Cordapp-Contract-Version";
    static final String CORDAPP_WORKFLOW_NAME = "Cordapp-Workflow-Name";
    static final String CORDAPP_WORKFLOW_VERSION = "Cordapp-Workflow-Version";
    public static final String CORDA_CPK_TYPE = "Corda-CPK-Type";

    static final String CLASSES_IN_WRONG_DIRECTORY_FIXUP
        = "\"Classes found in the wrong directory\";"
            + FIXUPMESSAGES_RESTRICT_DIRECTIVE + '=' + FIXUPMESSAGES_IS_ERROR + ';'
            + FIXUPMESSAGES_IS_DIRECTIVE + '=' + FIXUPMESSAGES_IS_WARNING;

    /*
     * Location of official R3 documentation for building CPKs and CPBs.
     */
    static final String CORDAPP_DOCUMENTATION_URL = "https://docs.corda.net/cordapp-build-systems.html";

    static final String SEPARATOR = System.lineSeparator() + "- ";

    @NotNull
    static <T, R> List<R> map(@NotNull Iterable<T> items, @NotNull Function<T, R> operator) {
        return mapTo(new LinkedList<>(), items, operator);
    }

    @NotNull
    static <T, R, C extends Collection<R>> C mapTo(
        @NotNull C collection,
        @NotNull Iterable<T> items,
        @NotNull Function<T, R> operator
    ) {
        for (T item : items) {
            collection.add(operator.apply(item));
        }
        return collection;
    }

    @NotNull
    static <T, R, C extends Collection<R>> C flatMapTo(
        @NotNull C collection,
        @NotNull Iterable<T> items,
        @NotNull Function<T, ? extends Iterable<R>> transform
    ) {
        for (T item : items) {
            transform.apply(item).iterator().forEachRemaining(collection::add);
        }
        return collection;
    }

    @NotNull
    static <T, R> List<R> filterIsInstance(@NotNull Iterable<T> items, @NotNull Class<R> clazz) {
        final List<R> result = new LinkedList<>();
        for (T item : items) {
            if (clazz.isInstance(item)) {
                result.add(clazz.cast(item));
            }
        }
        return result;
    }

    @NotNull
    static String toMaven(@NotNull Dependency dependency) {
        final StringBuilder builder = new StringBuilder();
        final String group = dependency.getGroup();
        if (group != null) {
            builder.append(group).append(':');
        }
        builder.append(dependency.getName());
        final String version = dependency.getVersion();
        if (version != null) {
            builder.append(':').append(version);
        }
        return builder.toString();
    }

    @NotNull
    static Configuration createBasicConfiguration(
        @NotNull ConfigurationContainer container,
        @NotNull String name
    ) {
        final Configuration configuration = container.maybeCreate(name).setVisible(false);
        configuration.setCanBeConsumed(false);
        configuration.setCanBeResolved(false);
        return configuration;
    }

    /*
     * This configuration will contribute to the CorDapp's compile classpath
     * but not to its runtime classpath. However, it will still contribute
     * to all TESTING classpaths.
     */
    @NotNull
    static Configuration createCompileConfiguration(@NotNull ConfigurationContainer container, @NotNull String name) {
        return createCompileConfiguration(container, name, "Implementation");
    }

    @SuppressWarnings("SameParameterValue")
    @NotNull
    private static Configuration createCompileConfiguration(
        @NotNull ConfigurationContainer container,
        @NotNull String name,
        @NotNull String testSuffix
    ) {
        final Configuration configuration = createBasicConfiguration(container, name);
        container.getByName(COMPILE_ONLY_CONFIGURATION_NAME).extendsFrom(configuration);
        container.matching(cfg -> cfg.getName().endsWith(testSuffix)).configureEach(cfg ->
            cfg.extendsFrom(configuration)
        );
        return configuration;
    }

    // Configuration.setCanBeDeclared(boolean) was introduced in Gradle 8.2.
    public static void setCannotBeDeclared(@NotNull Configuration configuration) {
        try {
            MethodHandles.publicLookup()
                .findVirtual(configuration.getClass(), "setCanBeDeclared", methodType(void.class, boolean.class))
                .invoke(configuration, false);
        } catch (Error e) {
            throw e;
        } catch (Throwable ignored) {
        }
    }

    /**
     * Identify the artifacts that were resolved for these {@link Dependency} objects,
     * including all of their transitive dependencies.
     * @return The unique {@link ResolvedArtifact}s.
     */
    @NotNull
    static Set<ResolvedArtifact> resolveAll(
        @NotNull ResolvedConfiguration resolved,
        @NotNull Collection<? extends Dependency> dependencies
    ) {
        return resolve(resolved, dependencies, ResolvedDependency::getAllModuleArtifacts);
    }

    /**
     * Identify the artifacts that were resolved for these {@link Dependency} objects only.
     * This does not include any transitive dependencies.
     * @return The unique {@link ResolvedArtifact}s.
     */
    @NotNull
    static Set<ResolvedArtifact> resolveFirstLevel(
        @NotNull ResolvedConfiguration resolved,
        @NotNull Collection<? extends Dependency> dependencies
    ) {
        return resolve(resolved, dependencies, ResolvedDependency::getModuleArtifacts);
    }

    @NotNull
    private static Set<ResolvedArtifact> resolve(
        @NotNull ResolvedConfiguration resolved,
        @NotNull Collection<? extends Dependency> dependencies,
        @NotNull Function<ResolvedDependency, Iterable<ResolvedArtifact>> fetchArtifacts
    ) {
        return flatMapTo(
            new LinkedHashSet<>(),
            resolved.getFirstLevelModuleDependencies(dependencies::contains),
            fetchArtifacts
        );
    }

    @NotNull
    static <T> Set<T> subtract(@NotNull Set<T> originals, @NotNull Collection<T> unwanted) {
        final Set<T> result = new LinkedHashSet<>(originals);
        result.removeAll(unwanted);
        return result;
    }

    /**
     * @return The {@link Manifest} from a jar file.
     */
    @NotNull
    static Manifest manifestOf(@NotNull File file) {
        try (JarFile jarFile = new JarFile(file)) {
            return jarFile.getManifest();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    /**
     * @return The maximum value of {@code attributeName} from all {@link Manifest}s, or
     * {@code defaultValue} if no such value can be derived. The attribute is assumed to
     * have an integer value.
     */
    static int maxOf(@NotNull Iterable<File> files, @NotNull String attributeName, int defaultValue) {
        final Collection<Integer> candidates = new HashSet<>();
        for (File file : files) {
            final String value = manifestOf(file).getMainAttributes().getValue(attributeName);
            if (value != null) {
                try {
                    candidates.add(Integer.valueOf(value));
                } catch (NumberFormatException ignored) {
                }
            }
        }
        return candidates.isEmpty() ? defaultValue : max(candidates);
    }

    @NotNull
    static String joinToString(@NotNull Iterable<?> items, @NotNull CharSequence delimiter) {
        return joinToString(items, delimiter, "");
    }

    @NotNull
    static String joinToString(@NotNull Iterable<?> items, CharSequence delimiter, @NotNull CharSequence prefix) {
        final StringJoiner joiner = new StringJoiner(delimiter, prefix, "");
        for (Object item: items) {
            joiner.add(item.toString());
        }
        return joiner.toString();
    }

    @NotNull
    public static Action<TaskExecutionGraph> copyJarEnabledTo(@NotNull Task target) {
        return new CopyEnabledAction(target, Jar.class, JAR_TASK_NAME);
    }

    private static final Pattern MULTI_RELEASE = Pattern.compile("^META-INF/versions/\\d++/(.++)$");
    private static final char DIRECTORY_SEPARATOR = '/';
    private static final char PACKAGE_SEPARATOR = '.';

    /**
     * @return The package names from a {@link JarFile}.
     */
    @NotNull
    static Set<String> getPackages(@NotNull File file) {
        try (JarFile jar = new JarFile(file)) {
            final Set<String> packages = new LinkedHashSet<>();
            final Enumeration<JarEntry> jarEntries = jar.entries();
            while (jarEntries.hasMoreElements()) {
                final JarEntry jarEntry = jarEntries.nextElement();
                if (!jarEntry.isDirectory()) {
                    final String entryName = jarEntry.getName();
                    if (entryName.startsWith("OSGI-INF/")) {
                        continue;
                    }

                    final String binaryFQN;
                    if (entryName.startsWith("META-INF/")) {
                        final Matcher matcher = MULTI_RELEASE.matcher(entryName);
                        if (!matcher.matches()) {
                            continue;
                        }
                        binaryFQN = matcher.group(1);
                    } else {
                        binaryFQN = entryName;
                    }
                    final int lastIdx = binaryFQN.lastIndexOf(DIRECTORY_SEPARATOR);
                    final String binaryPackageName = lastIdx == -1 ? binaryFQN : binaryFQN.substring(0, lastIdx);
                    if (isValidPackage(binaryPackageName)) {
                        packages.add(toPackageName(binaryPackageName));
                    }
                }
            }
            return packages;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private static boolean isValidPackage(@NotNull String name) {
        return Arrays.stream(name.split(String.valueOf(DIRECTORY_SEPARATOR)))
            .allMatch(CordappUtils::isJavaIdentifier);
    }

    @NotNull
    private static String toPackageName(@NotNull String resourceName) {
        return resourceName.replace(DIRECTORY_SEPARATOR, PACKAGE_SEPARATOR);
    }

    /**
     * Convert a {@link String} into a {@link Map} of 'key,value' pairs.
     * Any lines which do not match the patterns 'key = value'
     * or 'key: value' are skipped.
     * @return The set of {@code (key, value)} pairs.
     */
    @NotNull
    static Map<String, String> parseInstructions(@NotNull String instructions) {
        final Map<String, String> result = new LinkedHashMap<>();
        new BufferedReader(new StringReader(instructions)).lines()
            .map(String::trim)
            .filter(s -> !s.isEmpty())
            .forEach(line -> {
                final String[] instruction = parseInstruction(line);
                if (instruction != null) {
                    result.put(instruction[0], instruction[1]);
                }
            });
        return unmodifiableMap(result);
    }

    @Nullable
    static String[] parseInstruction(@NotNull String instruction) {
        for (int idx = 0; idx < instruction.length(); ++idx) {
            final char c = instruction.charAt(idx);
            if (c == '=' || c == ':') {
                final String key = instruction.substring(0, idx).trim();
                return key.isEmpty() ? null : new String[] {
                    key, instruction.substring(idx + 1).trim()
                };
            }
        }
        return null;
    }

    /**
     * Check whether every {@link String} in this {@link List} is
     * a valid Java identifier.
     * @return true if all members of {@code strs} is a valid Java identifier.
     */
    static boolean isJavaIdentifiers(@NotNull List<String> strs) {
        return strs.stream().allMatch(CordappUtils::isJavaIdentifier);
    }

    /**
     * Checks whether this {@link String} could be considered to
     * be a valid identifier in Java. Identifiers are only
     * permitted to contain a specific subset of {@code char}s.
     * @return true if {@code str} is a valid Java identifier.
     */
    static boolean isJavaIdentifier(@NotNull String str) {
        if (str.isEmpty() || !Character.isJavaIdentifierStart(str.charAt(0))) {
            return false;
        }

        int idx = str.length();
        while (--idx > 0) {
            if (!Character.isJavaIdentifierPart(str.charAt(idx))) {
                return false;
            }
        }

        return true;
    }

    /**
     * @param algorithmName The hashing algorithm to use.
     * @return A {@link MessageDigest} for {@code algorithmName},
     * handling any {@link NoSuchAlgorithmException}.
     */
    @NotNull
    public static MessageDigest digestFor(@NotNull String algorithmName) {
        try {
            return MessageDigest.getInstance(algorithmName);
        } catch (NoSuchAlgorithmException e) {
            throw new InvalidUserDataException("Hash algorithm " + algorithmName + " not available");
        }
    }

    private static final int DEFAULT_BUFFER_SIZE = 8192;
    private static final int EOF = -1;

    /*
     * @return Hash for contents of {@link InputStream}.
     */
    @NotNull
    public static byte[] hashFor(@NotNull MessageDigest digest, @NotNull InputStream input) throws IOException {
        final byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
        while (true) {
            final int length = input.read(buffer);
            if (length == EOF) {
                break;
            }
            digest.update(buffer, 0, length);
        }
        return digest.digest();
    }

    static boolean isNullOrEmpty(@Nullable String s) {
        return s == null || s.isEmpty();
    }

    @NotNull
    static Element appendElement(@NotNull Element element, @NotNull String name) {
        final Element childElement = element.getOwnerDocument().createElement(name);
        element.appendChild(childElement);
        return childElement;
    }

    @NotNull
    static Element appendElement(@NotNull Element element, @NotNull String name, @Nullable String value) {
        final Element child = appendElement(element, name);
        child.appendChild(element.getOwnerDocument().createTextNode(value));
        return child;
    }
}

File: cordapp-cpk2/src/main/kotlin/net/corda/plugins/cpk2/CordappUtils.kt
@file:JvmName("CordappUtils")
package net.corda.plugins.cpk2

import net.corda.plugins.cpk2.XMLFactory.createDocumentBuilderFactory
import net.corda.plugins.cpk2.XMLFactory.createTransformerFactory
import org.gradle.api.InvalidUserDataException
import org.gradle.api.artifacts.Configuration
import org.gradle.api.artifacts.ConfigurationContainer
import org.gradle.api.artifacts.Dependency
import org.gradle.api.artifacts.ResolvedArtifact
import org.gradle.api.artifacts.ResolvedConfiguration
import org.gradle.api.artifacts.ResolvedDependency
import org.gradle.api.plugins.JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME
import org.gradle.api.specs.Spec
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.Writer
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.jar.JarFile
import java.util.jar.Manifest
import javax.xml.transform.OutputKeys.ENCODING
import javax.xml.transform.OutputKeys.INDENT
import javax.xml.transform.OutputKeys.METHOD
import javax.xml.transform.OutputKeys.OMIT_XML_DECLARATION
import javax.xml.transform.OutputKeys.STANDALONE
import javax.xml.transform.TransformerException
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

const val CORDAPP_CPK_PLUGIN_ID = "net.corda.plugins.cordapp-cpk2"
const val CORDAPP_TASK_GROUP = "Cordapp"
const val CPK_XML_NAMESPACE = "urn:corda-cpk"

const val CORDA_API_GROUP = "net.corda"
const val ENTERPRISE_API_GROUP = "com.r3.corda"

const val CORDAPP_SEALING_SYSTEM_PROPERTY_NAME = "net.corda.cordapp.sealing.enabled"

const val CPK_FILE_EXTENSION = "jar"
const val CPK_ARTIFACT_CLASSIFIER = "cordapp"

const val CORDAPP_CONFIGURATION_NAME = "cordapp"
const val CORDAPP_EXTERNAL_CONFIGURATION_NAME = "cordappExternal"
const val CORDAPP_PACKAGING_CONFIGURATION_NAME = "cordappPackaging"
const val CORDA_RUNTIME_ONLY_CONFIGURATION_NAME = "cordaRuntimeOnly"
const val CORDA_ALL_PROVIDED_CONFIGURATION_NAME = "cordaAllProvided"
const val CORDA_PRIVATE_CONFIGURATION_NAME = "cordaPrivateProvided"
const val CORDA_PROVIDED_CONFIGURATION_NAME = "cordaProvided"
const val CORDA_EMBEDDED_CONFIGURATION_NAME = "cordaEmbedded"
const val ALL_CORDAPPS_CONFIGURATION_NAME = "allCordapps"
const val CORDA_CPK_CONFIGURATION_NAME = "cordaCPK"

const val CPK_DEPENDENCIES = "META-INF/CPKDependencies"

// These tags are for the CPK file.
const val CPK_PLATFORM_VERSION = "Corda-CPK-Built-Platform-Version"
const val CPK_CORDAPP_NAME = "Corda-CPK-Cordapp-Name"
const val CPK_CORDAPP_VERSION = "Corda-CPK-Cordapp-Version"
const val CPK_CORDAPP_LICENCE = "Corda-CPK-Cordapp-Licence"
const val CPK_CORDAPP_VENDOR = "Corda-CPK-Cordapp-Vendor"
const val CPK_FORMAT_TAG = "Corda-CPK-Format"
const val CPK_FORMAT = "2.0"

// These tags are for the "main" JAR file.
const val PLATFORM_VERSION_X = 999
const val CORDA_CONTRACT_CLASSES = "Corda-Contract-Classes"
const val CORDA_WORKFLOW_CLASSES = "Corda-Flow-Classes"
const val CORDA_MAPPED_SCHEMA_CLASSES = "Corda-MappedSchema-Classes"
const val CORDA_SERVICE_CLASSES = "Corda-Service-Classes"
const val IMPORT_POLICY_PACKAGES = "Import-Policy-Packages"
const val REQUIRED_PACKAGES = "Required-Packages"

const val CORDAPP_PLATFORM_VERSION = "Cordapp-Built-Platform-Version"
const val CORDAPP_CONTRACT_NAME = "Cordapp-Contract-Name"
const val CORDAPP_CONTRACT_VERSION = "Cordapp-Contract-Version"
const val CORDAPP_WORKFLOW_NAME = "Cordapp-Workflow-Name"
const val CORDAPP_WORKFLOW_VERSION = "Cordapp-Workflow-Version"
const val CORDA_CPK_TYPE = "Corda-CPK-Type"

/**
 * Location of official R3 documentation for building CPKs and CPBs.
 */
const val CORDAPP_DOCUMENTATION_URL = "https://docs.corda.net/cordapp-build-systems.html"

@JvmField
val SEPARATOR: String = System.lineSeparator() + "- "

fun Dependency.toMaven(): String {
    val builder = StringBuilder()
    group?.also { builder.append(it).append(':') }
    builder.append(name)
    version?.also { builder.append(':').append(it) }
    return builder.toString()
}

fun ConfigurationContainer.createBasicConfiguration(name: String): Configuration {
    return maybeCreate(name)
        .setVisible(false)
        .also { configuration ->
            configuration.isCanBeConsumed = false
            configuration.isCanBeResolved = false
        }
}

/**
 * This configuration will contribute to the CorDapp's compile classpath
 * but not to its runtime classpath. However, it will still contribute
 * to all TESTING classpaths.
 */
fun ConfigurationContainer.createCompileConfiguration(name: String): Configuration {
    return createCompileConfiguration(name, "Implementation")
}

private fun ConfigurationContainer.createCompileConfiguration(name: String, testSuffix: String): Configuration {
    return createBasicConfiguration(name).also { configuration ->
        getByName(COMPILE_ONLY_CONFIGURATION_NAME).extendsFrom(configuration)
        matching { it.name.endsWith(testSuffix) }.configureEach { cfg ->
            cfg.extendsFrom(configuration)
        }
    }
}

/**
 * Identify the artifacts that were resolved for these [Dependency] objects,
 * including all of their transitive dependencies.
 */
fun ResolvedConfiguration.resolveAll(dependencies: Collection<Dependency>): Set<ResolvedArtifact> {
    return resolve(dependencies, ResolvedDependency::getAllModuleArtifacts)
}

/**
 * Identify the artifacts that were resolved for these [Dependency] objects only.
 * This does not include any transitive dependencies.
 */
fun ResolvedConfiguration.resolveFirstLevel(dependencies: Collection<Dependency>): Set<ResolvedArtifact> {
    return resolve(dependencies, ResolvedDependency::getModuleArtifacts)
}

private fun ResolvedConfiguration.resolve(dependencies: Collection<Dependency>, fetchArtifacts: (ResolvedDependency) -> Iterable<ResolvedArtifact>): Set<ResolvedArtifact> {
    return getFirstLevelModuleDependencies(Spec(dependencies::contains))
        .flatMapTo(LinkedHashSet(), fetchArtifacts)
}

/**
 * Extracts the [Manifest] from a jar file.
 */
val File.manifest: Manifest get() = JarFile(this).use(JarFile::getManifest)

/**
 * Computes the maximum value of [attributeName] from all [Manifest]s,
 * or `null` if no such value can be derived. The attribute is assumed
 * to have an integer value.
 */
fun Iterable<File>.maxOf(attributeName: String): Int? {
    return mapNotNull { it.manifest.mainAttributes.getValue(attributeName)?.toIntOrNull() }.maxOrNull()
}

/**
 * Extracts the package names from a [JarFile].
 */
private val MULTI_RELEASE = "^META-INF/versions/\\d++/(.++)\$".toRegex()
private const val DIRECTORY_SEPARATOR = '/'
private const val PACKAGE_SEPARATOR = '.'

val File.packages: MutableSet<String> get() {
    return JarFile(this).use { jar ->
        val packages = mutableSetOf<String>()
        val jarEntries = jar.entries()
        while (jarEntries.hasMoreElements()) {
            val jarEntry = jarEntries.nextElement()
            if (!jarEntry.isDirectory) {
                val entryName = jarEntry.name
                if (entryName.startsWith("OSGI-INF/")) {
                    continue
                }

                val binaryFQN = if (entryName.startsWith("META-INF/")) {
                    (MULTI_RELEASE.matchEntire(entryName) ?: continue).groupValues[1]
                } else {
                    entryName
                }
                val binaryPackageName = binaryFQN.substringBeforeLast(DIRECTORY_SEPARATOR)
                if (isValidPackage(binaryPackageName)) {
                    packages.add(binaryPackageName.toPackageName())
                }
            }
        }
        packages
    }
}

private fun isValidPackage(name: String): Boolean {
    return name.split(DIRECTORY_SEPARATOR).all(String::isJavaIdentifier)
}

private fun String.toPackageName(): String {
    return replace(DIRECTORY_SEPARATOR, PACKAGE_SEPARATOR)
}

/**
 * Check whether every [String] in this [List] is
 * a valid Java identifier.
 */
val List<String>.isJavaIdentifiers: Boolean get() {
    return this.all(String::isJavaIdentifier)
}

/**
 * Checks whether this [String] could be considered to
 * be a valid identifier in Java. Identifiers are only
 * permitted to contain a specific subset of [Character]s.
 */
val String.isJavaIdentifier: Boolean get() {
    if (isEmpty() || !Character.isJavaIdentifierStart(this[0])) {
        return false
    }

    var idx = length
    while (--idx > 0) {
        if (!Character.isJavaIdentifierPart(this[idx])) {
            return false
        }
    }

    return true
}

/**
 * Get a [MessageDigest] for [algorithmName], and handle
 * any [NoSuchAlgorithmException].
 */
fun digestFor(algorithmName: String): MessageDigest {
    return try {
        MessageDigest.getInstance(algorithmName)
    } catch (_ : NoSuchAlgorithmException) {
        throw InvalidUserDataException("Hash algorithm $algorithmName not available")
    }
}

private const val EOF = -1

/**
 * Compute hash for contents of [InputStream].
 */
@Throws(IOException::class)
fun MessageDigest.hashFor(input: InputStream): ByteArray {
    val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
    while (true) {
        val length = input.read(buffer)
        if (length == EOF) {
            break
        }
        update(buffer, 0, length)
    }
    return digest()
}

/**
 * Helper functions for XML documents.
 * Note that creating factories is EXPENSIVE.
 */
private val documentBuilderFactory = createDocumentBuilderFactory()

fun createXmlDocument(): Document {
    return documentBuilderFactory.newDocumentBuilder().newDocument().apply {
        xmlStandalone = true
    }
}

private val transformerFactory = createTransformerFactory()

@Throws(TransformerException::class)
fun Document.writeTo(writer: Writer) {
    val transformer = transformerFactory.newTransformer()
    transformer.setOutputProperty(METHOD, "xml")
    transformer.setOutputProperty(ENCODING, "UTF-8")
    transformer.setOutputProperty(INDENT, "yes")
    transformer.setOutputProperty(STANDALONE, "yes")
    transformer.setOutputProperty(OMIT_XML_DECLARATION, "no")
    transformer.transform(DOMSource(this), StreamResult(writer))
}

fun Document.createRootElement(namespace: String, name: String): Element {
    val rootElement = createElementNS(namespace, name)
    appendChild(rootElement)
    return rootElement
}

fun Element.appendElement(name: String): Element {
    val childElement = ownerDocument.createElement(name)
    appendChild(childElement)
    return childElement
}

fun Element.appendElement(name: String, value: String?): Element {
    return appendElement(name).also { child ->
        child.appendChild(ownerDocument.createTextNode(value))
    }
}
